apps/nextjs/src/stores/lessonPlanStore/stateActionFunctions/handleMessagesUpdated.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  applyLessonPlanPatchImmutable,
+  extractPatches,
+  PatchDocument,
+} from "@oakai/aila/src/protocol/jsonPatchProtocol";
+import { aiLogger } from "@oakai/logger";
+import { createHash } from "crypto";
+
+import type { AiMessage } from "@/stores/chatStore/types";
+
+import type { LessonPlanStore } from "..";
+
+const log = aiLogger("lessons:store");
+
+const timeOperation = <T>(fn: () => T): T => {
+  const startTime = performance.now();
+  const result = fn();
+  const endTime = performance.now();
+  log.info(`applyLessonPlanPatch took ${endTime - startTime} milliseconds`);
+  return result;
+};
+
+const generatePatchHash = (patch: PatchDocument): string => {
+  const patchString = JSON.stringify(patch);
+  const hash = createHash("sha256");
+  hash.update(patchString);
+  return hash.digest("hex");
+};
+
+const countCompleteParts = (messageStr: string) => {
+  return [...messageStr.matchAll(/"status":"complete"/g)].length;
+};
+
+export const handleMessagesUpdated =
+  (
+    set: (partial: Partial<LessonPlanStore>) => void,
+    get: () => LessonPlanStore,
+  ) =>
+  (messages: AiMessage[], isLoading: boolean) => {
+    if (!get().isAcceptingChanges) {
+      log.info("message skipped: not accepting changes");
+      return;
+    }
+    const streamingMessage = messages[messages.length - 1];
+    if (!streamingMessage) {
+      log.info("message skipped: no message");
+      return;
+    }
+    if (streamingMessage.role !== "assistant") {
+      log.info("message skipped: wrong role");
+      return;
+    }
+
+    const numberOfStreamedCompleteParts = countCompleteParts(
+      streamingMessage.content,
+    );
```

Comment:
The streaming message is constantly changing while streaming, but we don't need to recalculate partial parches while they stream, just the patches that have finished streaming

We know that all patches finish with a `"status":"complete"` string, so we can count this with string logic to determine if a new patch is available

Because we only apply patches when they're complete, we can use string logic to count the number of `"status": "complete"` strings 
The only thing we need to know is whether a